File: bank_app/services/customer_service.py
from repositories.customer_repository import customer_repository

class CustomerService:
    def create_customer(self, name, email, branch_id):
        return customer_repository.insert({"name": name, "email": email, "branch_id": branch_id})

    def list_customers(self):
        return customer_repository.find_all()
    
    def get_customer(self, customer_id):
        return customer_repository.find_by_id(customer_id)

    def delete_customer(self, customer_id):
        return customer_repository.delete(customer_id)

customer_service = CustomerService()

File: bank-ui/src/services/customerService.ts
export type Customer = {
    id: string
    name: string
    email: string
    branch_id: string
}

const API_URL = 'http://127.0.0.1:8000/api/v1/customers'

export const fetchCustomers = async (): Promise<Customer[]> => {
    const response = await fetch(API_URL)
    if (!response.ok) {
        throw new Error(`Failed to fetch customers (${response.status})`)
    }
    return response.json()
}

export const createCustomer = async (name: string, email: string, branchId: string): Promise<Customer> => {
    const response = await fetch(API_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ name: name, email: email, branch_id: branchId })
    })
    if (!response.ok) {
        throw new Error(`Failed to create customer (${response.status})`)
    }
    return response.json()
}

export const fetchCustomerById = async (customerId: string): Promise<Customer> => {
    const response = await fetch(`${API_URL}/${customerId}`)
    if (!response.ok) {
        throw new Error(`Failed to find customer (${response.status})`)
    }
    return response.json()
}

export const deleteCustomer = async (customerId: string): Promise<void> => {
    const response = await fetch(`${API_URL}/${customerId}`, {
        method: 'DELETE',
    })
    if (!response.ok) {
        throw new Error(`Failed to find customer (${response.status})`)
    }
}